### PR TITLE
bazel: add nogo and define bodyclose linter locally

### PR DESCRIPTION
**Pull request by burmudar**
This adds nogo to lint our code from within bazel.
## Test plan
`bazel build //cmd/frontend`

need to do a full build still

### CODE REVIEW

File: BUILD.bazel [hunk 1]
```diff
@@ -195,4 +195,27 @@ js_library(
     ],
 )
 
+# Go
+
+# nogo config
+#
+# For nogo to be able to run a linter, it needs to have `var Analyzer analysis.Analyzer` defined in the main package.
```
Comment (jhchabran) [inline BUILD.bazel:202]: Ah! It needs to be in the main package, I missed that on my first read of the docs. That explains a lot.

File: WORKSPACE [hunk 2]
```diff
@@ -193,6 +193,7 @@ rules_proto_toolchains()
 load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")
 load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")
 load("//:deps.bzl", "go_dependencies")
+load("//:linters.bzl", "linter_dependencies")
```
Comment (jhchabran) [inline WORKSPACE:196]:
```suggestion
load("//:linters_deps.bzl", "linter_dependencies")
```

A bit nitpicky, but I think that will help if you don't mind renaming this.

File: dev/linters/bodyclose/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    srcs = ["bodyclose.go"],
+    importpath = "github.com/sourcegraph/sourcegraph/dev/linters/bodyclose",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@com_github_timakin_bodyclose//passes/bodyclose",
+        "@org_golang_x_tools//go/analysis:go_default_library",
+    ],
+)
+
+#exports_files(["nogo_config.json"])
```
Comment (jhchabran) [inline dev/linters/bodyclose/BUILD.bazel:14]: left over

File: linters.bzl [hunk 4]
```diff
@@ -0,0 +1,13 @@
+load("@bazel_gazelle//:deps.bzl", "go_repository")
+
+def linter_dependencies():
+  go_repository(
+    name = "com_github_timakin_bodyclose",
+    build_file_proto_mode = "disable_global",
+    importpath = "github.com/timakin/bodyclose",
+    # NOTE: Gazelle will not generate this for you
+    # To retrieve these values you can create a go.mod in this directory
+    # and then run `go mod tidy` and delete the go.mod you created.
```
Comment (jhchabran) [inline linters.bzl:10]: That's pretty classic Bazel, so I think this should be a FAQ entry instead.

File: nogo_config.json [hunk 5]
```diff
@@ -0,0 +1,10 @@
+{
+    "bodyclose": {
+    "exclude_files": {
+      "external/": "no need to vet third party code",
+      ".*_generated\\.go$": "ignore generated code",
```
Comment (jhchabran) [inline nogo_config.json:5]: I think we should exclude `*.pb.go` as well then!

File: go.sum [hunk 6]
```diff
@@ -2200,6 +2201,7 @@ github.com/tidwall/pretty v1.0.0/go.mod h1:XNkn88O1ChpSDQmQeStsy+sBenx6DDtFZJxhV
 github.com/tidwall/pretty v1.2.0 h1:RWIZEg2iJ8/g6fDDYzMpobmaoGh5OLl4AXtGUGPcqCs=
 github.com/tidwall/pretty v1.2.0/go.mod h1:ITEVvHYasfjBbM0u2Pg8T2nJnzm8xPwvNhhsoaGGjNU=
 github.com/timakin/bodyclose v0.0.0-20190930140734-f7f2e9bca95e/go.mod h1:Qimiffbc6q9tBWlVV6x0P9sat/ao1xEkREYPPj9hphk=
+github.com/timakin/bodyclose v0.0.0-20200424151742-cb6215831a94 h1:ig99OeTyDwQWhPe2iw9lwfQVF1KB3Q4fpP3X7/2VBG8=
```
Comment (jhchabran) [inline go.sum:2204]: We don't want to have this in our go.mod, other wise it's highly possible that this gets deleted in a PR where someone ran go mod tidy I think.

Comment (jhchabran) [reply go.sum:2204]: EDIT: oh right, we've been including this in the linters folder, so we kinda still need this. Perhaps we could trick around with a go.mod in that folder? I'm a bit ambivalent on this, in principle it would be better to avoid pulling deps we don't need to build our binaries, but this might very well have practically zero deps too.

**Review by jhchabran — COMMENTED** [review]
It's great to see that we can tackle this with less issues than expected, this is going to speed up PRs by a lot :)

File: WORKSPACE (hunk 2 above)
Comment (burmudar) [reply WORKSPACE:196]: I agree completely actually.

TL;DR: Bazel actually recommends that file targets be colocated where they are used, meaning, since `nogo_config.json` is used in the root BUILD.bazel in the `nogo` rule, it's better to have the file in the root and to refer to it as `//:nogo_config.json`

 I did go down a mini rabbit hole in trying to have all of this co-located in the //dev/linters package but I had trouble referencing the `nogo_config.json` file in the repo root BUILD.bazel. 

`bazel query` showed the target existed:
```
bazel query '//dev/linters:*'
//dev/linters:BUILD.bazel
//dev/linters:nogo_config.json
```
`bazel cquery //dev/linters:nogo_config.json --output=files` showed the file:
```
bazel cquery //dev/linters:nogo_config.json --output=files
INFO: Analyzed target //dev/linters:nogo_config.json (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
dev/linters/nogo_config.json
INFO: Elapsed time: 0.236s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 0 total actions
```

but once you refer to the file in the config attribute of the nogo rule, you would get:
```
bazel build //cmd/frontend
ERROR: /Users/william/code/sourcegraph/BUILD.bazel:205:5: no such target '//dev/linters:nogo_config.json)': target 'nogo_config.json)' not declared in package 'dev/linters' defined by /Users/william/code/sourcegraph/dev/linters/BUILD.bazel (did you mean 'nogo_config.json'? Tip: use `query "//dev/linters:*"` to see all the targets in that package) and referenced by '//:sg_nogo_actual'
```

I also tried `file_group` and referring to the location with `$(location )` but no 🎲

File: linters.bzl (hunk 4 above)
Comment (burmudar) [reply linters.bzl:10]: cool - I'll move it

File: nogo_config.json (hunk 5 above)
Comment (burmudar) [reply nogo_config.json:5]: Ah good spot!

File: go.sum (hunk 6 above)
Comment (burmudar) [reply go.sum:2204]: I agree, I think we should go with the `go.mod` route - it's easily reversible. There is no guarantee once I have all the linters in that some linter pulls in some weird dependency and this way we can keep it isolated

**Review by jhchabran — APPROVED** [review]
Well done!
